api/src/pcapi/routes/auth/discord.py

```diff
@@ -103,23 +113,35 @@ def update_discord_user(user_id: str, discord_id: str) -> None:
     if already_linked_user:
         raise auth_exceptions.DiscordUserAlreadyLinked()
 
-    user = user_models.User.query.get(user_id)
+    user: user_models.User = user_models.User.query.get(user_id)
     discord_user = user.discordUser
 
     if discord_user is None:
-        # We still add the user to the database even if he doesn't have access to the discord server
         discord_user = user_models.DiscordUser(userId=user.id, discordId=discord_id, hasAccess=False)
-        db.session.add(discord_user)
-        raise auth_exceptions.UserNotAllowed()
+
+    if user.has_beneficiary_role:
+        discord_user.hasAccess = True
+    else:
+        discord_user.hasAccess = False
```

Comment:
Détail : ça me semble assez équivalent à cette version plus simple : 
```suggestion
    discord_user.hasAccess = user.has_beneficiary_role
```